File: frontend/src/services/loanService.js
import apiClient from './apiClient';
import { API_ENDPOINTS } from '../constants/apiEndpoints';

const loanService = {
  submitLoanApplication: async (loanApplicationData) => {
    // Safety check: if this prints undefined, your constants file is the problem
    console.log("Target URL:", API_ENDPOINTS?.LOAN_ENDPOINTS?.APPLY_LOAN);
    
    const response = await apiClient.post(
      API_ENDPOINTS.LOAN_ENDPOINTS.APPLY_LOAN, 
      loanApplicationData
    );
    return response;
  },

  /**
   * Get loan by ID
   */
  getLoanById: (loanId) => {
    return apiClient.get(API_ENDPOINTS.LOAN_ENDPOINTS.GET_LOAN(loanId));
  },

  getUserLoans: async (userId, page = 0, size = 10) => {
    const response = await apiClient.get(
      API_ENDPOINTS.LOAN_ENDPOINTS.GET_USER_LOANS(userId),
      { params: { page, size } }
    );
    return response;
  },

  /**
   * Get all loans (paginated)
   */
  getAllLoans: (page = 0, size = 10) => {
    return apiClient.get(API_ENDPOINTS.LOAN_ENDPOINTS.GET_ALL_LOANS, {
      params: { page, size }
    });
  },

  /**
   * Approve a loan application
   */
  approveLoan: (loanId) => {
    return apiClient.post(
      API_ENDPOINTS.LOAN_ENDPOINTS.APPROVE_LOAN(loanId)
    );
  },

  /**
   * Reject a loan application
   */
  rejectLoan: (loanId) => {
    return apiClient.post(
      API_ENDPOINTS.LOAN_ENDPOINTS.REJECT_LOAN(loanId)
    );
  },

  /**
   * Get loan by number
   */
  getLoanByNumber: (loanNumber) => {
    return apiClient.get(
      API_ENDPOINTS.LOAN_ENDPOINTS.GET_LOAN_BY_NUMBER(loanNumber)
    );
  }
};

export default loanService;

File: backend/src/main/java/com/snlbanking/service/LoanService.java
package com.snlbanking.service;

import com.snlbanking.dto.LoanApplicationRequestDTO;
import com.snlbanking.dto.LoanApplicationResponseDTO;
import com.snlbanking.entity.Loan;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

/**
 * Loan Service Interface
 * Manages loan operations
 * 
 * @author SNL Banking Team
 * @version 1.0.0
 */
public interface LoanService {

    /**
     * Submit a loan application
     * @param loanApplicationRequestDTO The loan application request
     * @return LoanApplicationResponseDTO with application details
     */
    LoanApplicationResponseDTO submitLoanApplication(LoanApplicationRequestDTO loanApplicationRequestDTO);

    /**
     * Get loan by ID
     * @param loanId The loan ID
     * @return LoanApplicationResponseDTO
     */
    LoanApplicationResponseDTO getLoanById(Long loanId);

    /**
     * Get all loans for a user
     * @param userId The user ID
     * @param pageable Pagination info
     * @return Page of loans
     */
    Page<LoanApplicationResponseDTO> getUserLoans(Long userId, Pageable pageable);

    /**
     * Get all loans
     * @param pageable Pagination info
     * @return Page of loans
     */
    Page<LoanApplicationResponseDTO> getAllLoans(Pageable pageable);

    /**
     * Approve a loan application
     * @param loanId The loan ID
     */
    void approveLoan(Long loanId);

    /**
     * Reject a loan application
     * @param loanId The loan ID
     */
    void rejectLoan(Long loanId);

    /**
     * Get loan by number
     * @param loanNumber The loan number
     * @return LoanApplicationResponseDTO
     */
    LoanApplicationResponseDTO getLoanByNumber(String loanNumber);
}
